aggregator/election_mgr.go

```diff
@@ -0,0 +1,518 @@
+	if currState == newState {
+		if newState != LeaderState {
+			return
+		}
+		// We are in the middle of verifying leader and changing from leader to follower
+		// and we received an update that indicates the current instance is now the leader,
+		// therefore we cancel the change from leader to follower and instead set the
+		// election state to leader.
+		if mgr.cancelInProgressChange() {
+			// NB(xichen): if an in-progress change was cancelled, we need to explicitly set
+			// election state to leader in case the role changing goroutine managed to change
+			// the state to follower just before it was cancelled.
+			mgr.Lock()
+			mgr.electionState = LeaderState
+			mgr.Unlock()
+			mgr.notifyStateChange()
+		}
+		return
+	}
+
+	// Changing role from follower to leader.
+	if currState == FollowerState && newState == LeaderState {
+		mgr.Lock()
+		mgr.electionState = newState
+		mgr.Unlock()
+		mgr.notifyStateChange()
+		mgr.metrics.followerToLeader.Inc(1)
+		mgr.logger.Info("election state changed from follower to leader")
+		return
+	}
+
+	// NB(xichen): in case of a leader to follower transition, we need to be extra
+	// careful and should confirm the leader role has been actually claimed before
+	// setting the election state to follower. Otherwise, we might get into a situation
+	// where we resigned from the leader role but the other side (i.e., the old follower)
+	// is not claiming the leader role due to e.g., hardware/network issues and as such,
+	// neither instances will be flushing data downstream.
+	mgr.Lock()
+	defer mgr.Unlock()
+	if mgr.changeInProgress {
+		return
+	}
+	mgr.changeInProgress = true
+	ctx, cancelFn := context.WithCancel(context.Background())
+	mgr.changeCancelFn = cancelFn
+	mgr.changeWg.Add(1)
+	go mgr.leaderToFollower(ctx)
+}
+
+// cancelInProgressChange cancels in progress election state changes if any,
+// returning true if a change is cancelled and false otherwise.
+func (mgr *electionManager) cancelInProgressChange() bool {
+	mgr.Lock()
+	if !mgr.changeInProgress {
+		mgr.Unlock()
+		return false
+	}
+	cancelFn := mgr.changeCancelFn
+	mgr.changeCancelFn = nil
+	mgr.changeInProgress = false
+	mgr.Unlock()
+	cancelFn()
+	mgr.changeWg.Wait()
+	return true
+}
+
+func (mgr *electionManager) leaderToFollower(ctx context.Context) {
+	defer mgr.changeWg.Done()
+
+	continueFn := func(int) bool {
+		select {
+		case <-ctx.Done():
+			return false
+		default:
+			return true
+		}
+	}
+
+	// NB(xichen): majority of the verification logic below will be moved into the
+	// leadership API implementation to ensure the API user will receive a follower
+	// status only after this verification is performed.
+	mgr.changeRetrier.AttemptWhile(continueFn, func() error {
+		leader, err := mgr.leaderService.Leader(mgr.electionKey)
+		if err != nil {
+			mgr.metrics.leaderErrors.Inc(1)
+			mgr.logError("error determining the leader", err)
+			return err
+		}
+		if leader == mgr.leaderValue {
```

Comment:
Isn't there some read lock that should be acquired to test against `mgr.leaderValue`?